### PR TITLE
ChainConfig refactor

### PR DESCRIPTION
**Pull request by elmato**

### CODE REVIEW

File: silkworm/core/chain/config.cpp [hunk 1]
```diff
@@ -201,12 +214,14 @@ evmc_revision ChainConfig::revision(const BlockHeader& header) const noexcept {
         return determine_revision_by_block(header.number, header.timestamp);
     }
 
-    //TODO: read nonce from header / if header.number==0 return istambul;
+    // Genesis nonce does not contain version info. Default to istambul.
     if(header.number == 0) {
         return *_revision ? *_revision : EVMC_ISTANBUL;
```
Comment (elmato) [inline silkworm/core/chain/config.cpp:219]: I think this should be 
```c++
return _revision.has_value() ? *_revision : EVMC_ISTANBUL;
```

File: silkworm/core/chain/config.cpp [hunk 2]
```diff
@@ -124,42 +145,44 @@ std::optional<ChainConfig> ChainConfig::from_json(const nlohmann::json& json) no
...
-    read_json_config_member(json, "shanghaiTime", config.shanghai_time);
-    read_json_config_member(json, "cancunTime", config.cancun_time);
+    read_json_config_member(json, "mergeNetsplitBlock", config._merge_netsplit_block);
+    read_json_config_member(json, "shanghaiTime", config._shanghai_time);
+    read_json_config_member(json, "cancunTime", config._cancun_time);
 
     /* Note ! genesis_hash is purposely omitted. It must be loaded from db after the
      * effective genesis block has been persisted */
 
+    read_json_config_member(json, "revision", config._revision);
```
Comment (elmato) [inline silkworm/core/chain/config.cpp:184]: If we want to modify the genesis json to reflect the fact that we are no longer retrieving the EVMC revision based on block-height or time, we should use the EOSEVM version as a genesis configuration field instead.

File: silkworm/core/chain/config.cpp (hunk 1 above)
Comment (elmato) [reply silkworm/core/chain/config.cpp:219]: a possible `ChainConfig::revision`  implementation

```c++
evmc_revision ChainConfig::revision(const BlockHeader& header) const noexcept {
    if(protocol_rule_set != protocol::RuleSetType::kTrust) {
        return determine_revision_by_block(header.number, header.timestamp);
    }
    uint64_t evm_version = 0;
    if(header.number == 0) {
        evm_version = _eosevm_version.has_value() ? *_eosevm_version : 0;
    } else {
        evm_version = nonce_to_eos_evm_version(header.nonce)
    }
    return eos_evm_version_to_evmc_revision(evm_version);
}
```

File: silkworm/core/chain/config.cpp [hunk 3]
```diff
@@ -51,6 +65,13 @@ static inline void read_json_config_member(const nlohmann::json& json, const std
     }
 }
 
+static inline void read_json_config_member(const nlohmann::json& json, const std::string& key,
```
Comment (elmato) [inline silkworm/core/chain/config.cpp:68]: we should remove this specialization

**Review by yarkinwho — APPROVED** [review]
lgtm

File: silkworm/core/chain/config.hpp [hunk 4]
```diff
@@ -57,34 +58,124 @@ struct ChainConfig {
     //! \brief Returns the type of the (pre-Merge) protocol rule set
     protocol::RuleSetType protocol_rule_set{protocol::RuleSetType::kNoProof};
 
+    const std::optional<BlockNum>& homestead_block() const{
```
Comment (yarkinwho) [inline silkworm/core/chain/config.hpp:61]: Are those accessor really being used anywhere?

Comment (elmato) [reply silkworm/core/chain/config.hpp:61]: Yes, there are some places in the code where those accessors are called individually and we are throwing an exception if the ruleset used is not kTrust

**Pull request merged**